### PR TITLE
DSD-1770: FilterBarInline Component

### PR DESCRIPTION
**Pull request by jackiequach**
Fixes JIRA ticket [DSD-1770](https://jira.nypl.org/browse/DSD-1770).

## This PR does the following:

- Adds the FilterBarInline Component

## How has this been tested?

Locally on Storybook

## Accessibility concerns or updates

-

### Checklist:

- [ ] I have updated the Storybook documentation accordingly.
- [ ] I have added relevant accessibility documentation for this pull request.
- [ ] All new and existing tests passed.

### Front End Review:

- [ ] Review the Vercel preview deployment once it is ready.

[DSD-1770]: https://newyorkpubliclibrary.atlassian.net/browse/DSD-1770?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.tsx]: > Noticed when the page loads, the mobile view is displayed before the desktop view due to the Chakra's useMediaQuery hook.

Regarding this comment, this approached worked for me https://github.com/NYPL/nypl-design-system/pull/1599/files#diff-1885fe882eadb679d28e96b476db7c366b4edabf310cafdffc31160cad0d67c4

I think you can get away without having `useNYPLBreakpoints` and doing more responsive styles if you have a theme object file. That's why youre' seeing the mobile view first before the desktop styles kick in.

File: src/components/FilterBarInline/FilterBarInline.stories.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.stories.tsx:357]: Double nested render prop pattern. Does this feel or seem difficult to keep track of?

**Review by EdwinGuzman — COMMENTED** [review]
Even as a draft, this is looking great! Seems there's more documentation than component code so it feels complex as a UI wrapper, but glad to see multiple examples. 

One small issue I ran into was the focus seemed to jump in the example under the "Passing Children" section. I clicked on the "Accessibility" MultiSelect, then checkbox 1, and then the text input. The focus did not persist on the text input even though I clicked on it repeatedly. This went away when I typed but the focus ring was not visible. I can't replicate this anymore but wanted to point it out.

I also ran into the mobile rendering issue and removing the `useNYPLBreakpoints` hook should fix that, though you will need to move the logic as responsive styles for the component. I do see what you mean about needing the value for the children in the `renderChildren` function.

The other option I'm thinking is that this is meant to be the desktop component and the mobile equivalent is the "popup" component. Perhaps you can flip the logic around so that the initial view is desktop, then use the value from `useNYPLBreakpoints` to conditionally render the children. But do want @bigfishdesign13 to verify this.

The component as is does not need track of data, which is good, but I think it needs the companion `useFilterBar` hook in examples to show how consuming devs can use both the hook and the component together to manage state. Sorry, not sure if that was clear or not but it might change a bit of how this is approached, but if it was decided to not use it for this round then ignore the comment.

**Comment by bigfishdesign13** [comment]
> The other option I'm thinking is that this is meant to be the desktop component and the mobile equivalent is the "popup" component.

Yes, what you stated is correct.  While the `FilterBarInline` component is responsive and does offer mobile styles, NYPL will only use it for the desktop viewport.  The `FilterBarPopup` component will be used for mobile viewports.  If you need to flip the style logic to ensure that the desktop styles render firwst, that would be fine.

File: src/components/FilterBarInline/FilterBarInline.stories.tsx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.stories.tsx:357]: It was a bit hard to read, so I added a note to breakout the render prop functions in the docs.

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.tsx]: The initial load still displays the mobile styles for the `MultiSelectGroup` component, as that component will need to be updated to not use `useNYPLBreakpoints`.

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx]: nitpick and super minor: we write this as `Prerelease`

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx:101]: This is true but it will only happen when the value is passed down correctly in the render function, right? I think it's worth saying that this will happen when the `FilterBarInline` component's values are passed down appropriately but you can say something like `see the "Passing Children" section for more information...".

This is also true for the next "Note".

Comment (EdwinGuzman) [reply src/components/FilterBarInline/FilterBarInline.mdx:101]: Actually, sorry, the `isBlockElement` is specific to the `MultiSelectGroup` but I think the above is still applicable to width and layout for other components (like `CheckboxGroup` and `RadioGroup`) right?

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx]: How about reordering this so it first shows the source code and then the relevant UI?

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx:228]: You give really good full examples below, but I think a short snippet to explain the shape of this will be helpful. Something like
```
<FilterBarInline
  id="..."
  heading="..."
  renderChildren={({ isBlockElement, layout, width }) => {
    return (...)
  }}
/>
```

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx]: I think you can remove this section entirely. You can still mention this and link to the appropriate doc page in the section above near the "To control each of the child form elements...." sentence. At that point, you can say "`useMultiSelect` is an optional but recommended hook ...." and then link to the docs.

It follows the same pattern as "Find documentation on how to do this on each individual component page".

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx:256]: I really like the flow of this section, showing code and explaining it as it gets to the final component render.

File: src/components/FilterBarInline/FilterBarInline.stories.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.stories.tsx]: Not a problem but just funny you don't follow the recommendation you gave about splitting this up into smaller functions!

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.tsx]: typo: `useNYPLBreapoints`

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.tsx]: I think `SelectedItems` is a bit confusing here and in the args table. What do you think about renaming this `SelectedItems as MultiSelectItems`? Then it's clear that the `FilterBarItemsType` accepts data from the `MultiSelect` component. I think that might still be confusing and don't have a good answer for this.

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.tsx]: There a barely visible flicker when the component loads and I think it's because of this. It doesn't seem like there's an easy way around this though, right?

Comment (EdwinGuzman) [reply src/components/FilterBarInline/FilterBarInline.tsx]: Perhaps load the desktop styles by default so that, if there is a flicker, it happens on mobile where we have less traffic. This still doesn't really solve the issue though...

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Looks great! I have mostly doc-related comments.

I am concerned about the flicker issue however. Because those are not values set through responsive styles in Chakra and are prop values instead, I can see that being an issue...

I'm tempted to ask for an example that includes _all_ DS filter components since you listed them down in the component description: `CheckboxGroup`, `MultiSelectGroup`, `RadioGroup`, `Checkbox`, `DatePicker`, `MultiSelect`, `Radio`, `Select`, `Slider`, `TextInput`, and `Toggle`. I'm not sure if we really need this right now. What we can do is create that working example in Turbine with all filter components.

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (bigfishdesign13) [inline src/components/FilterBarInline/FilterBarInline.mdx:305]: What do you think about including in this example the code for the `FilterBarInline` component with the `onClear` prop set to `onClearFilterBar`?

Comment (bigfishdesign13) [reply src/components/FilterBarInline/FilterBarInline.mdx:305]: Ok, I made it further down the page and I see that there is a culminating code block showing everyting tied together in the `return` statement.  This makes more sense now.  For further clarification, can you add appropriate subheadings for each subsection in this block?

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (bigfishdesign13) [reply src/components/FilterBarInline/FilterBarInline.mdx]: I agree with Edwin.  In a way, we've demoted the significance of this hook, so linking to the `useMultiSelect` docs should be enough.

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (bigfishdesign13) [inline src/components/FilterBarInline/FilterBarInline.mdx:106]: Feel free to use the `Banner` component to add notes like this.

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (bigfishdesign13) [inline src/components/FilterBarInline/FilterBarInline.tsx]: Similar to what is done in `Banner`, the `Heading` element needs to be set to default size.  In this case, it should be set to `heading5`.  To go along with that, add a new `FilterBarInline Heading` section right after the `Accessibility` section.  The new section should more or less mirror what we have in the `Banner` docs.

https://nypl.github.io/nypl-design-system/reservoir/v3/?path=/docs/components-feedback-banner--docs#banner-heading

**Review by bigfishdesign13 — CHANGES_REQUESTED** [review]
This is wonderful!  There are just a few things that need to be addressed.

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.mdx:101]: `isBlockElement` and `width` are specific to `MultiSelectGroup` and `layout` applies to the other components.

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.tsx]: I updated the component to default on desktop styles, but there is still a flicker that is caused by the MultiSelectGroup component.

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.mdx:106]: The Banner component would require special styling when using code snippets.

**Review by EdwinGuzman — APPROVED** [review]
Works well and the updates look great. We do need others to review before this gets merged in but you can still pass the Vercel preview URL to Clare for an accessibility review.

Not a concern but are we recommending devs/users to use this component with the "Apply filters" and "Clear all filters" buttons? I think it makes sense to recommend this pattern since that's a big reason to use this component. What do you think @bigfishdesign13 @jackiequach ?

**Comment by jackiequach** [comment]
> Not a concern but are we recommending devs/users to use this component with the "Apply filters" and "Clear all filters" buttons? I think it makes sense to recommend this pattern since that's a big reason to use this component. What do you think @bigfishdesign13 @jackiequach ?

@EdwinGuzman I agree, without the buttons it just seems like a container for form elements. I'll add the recommendation to the docs if Marty agrees.

**Comment by bigfishdesign13** [comment]
> Not a concern but are we recommending devs/users to use this component with the "Apply filters" and "Clear all filters" buttons? I think it makes sense to recommend this pattern since that's a big reason to use this component. What do you think @bigfishdesign13 @jackiequach ?

@jackiequach and @EdwinGuzman – The official UX recommendation is to use neither the Apply button nor the Clear button.  But rather to use real-time filtering (no need for Apply button) in conjunction with the TagSet component (includes a Clear button).  That said, we do want to offer the Apply and Clear buttons for apps that are not able to support the full UX recommendations right off the bat.

File: src/components/FilterBarInline/FilterBarInline.tsx
Comment (7emansell) [inline src/components/FilterBarInline/FilterBarInline.tsx]: I think it could be clearer here that using these buttons is not the officially recommended pattern (if that's the final call)

**Review by 7emansell — APPROVED** [review]
Looks fantastic to me!! just wanted a little more explicit clarification about dev usage

**Comment by jackiequach** [comment]
I added an example for using the `TagSet` component. It may be a bit complicated to implement when consuming apps using more filter components since the tag set data with the selected filter items. I think it might be better to add this to Turbine, but let me know your thoughts.

@bigfishdesign13 @EdwinGuzman @7emansell

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx:408]: Can you add a short explanation to say that the following code is just one example of how to use it but it could be done differently depending on the app's needs?

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (EdwinGuzman) [inline src/components/FilterBarInline/FilterBarInline.mdx:407]: Really good example. It feels a big laggy to me but do you notice it as well?

**Comment by EdwinGuzman** [comment]
> I added an example for using the TagSet component. It may be a bit complicated to implement when consuming apps using more filter components since the tag set data with the selected filter items. I think it might be better to add this to Turbine, but let me know your thoughts.

I think the text with the recommended approach and using the Tagset component is clear. I agree that a more complicated example should be shown in Turbine. One specific place to add this example with the TagSet component is in Turbine's Article and Databases page: https://nypl-ds-test-app.vercel.app/fullPages/articles-databases#above-header-notification

Right now, the MultiSelect components are commented out : https://github.com/NYPL/nypl-ds-test-app/blob/main/pages/fullPages/articles-databases.tsx#L142

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (jackiequach) [reply src/components/FilterBarInline/FilterBarInline.mdx:407]: I do notice some lag when updating the text input. The solution implemented for updating the tag set data is probably not the most efficient.

**Review by EdwinGuzman — APPROVED** [review]
Latest updates looks good to me!

File: src/components/FilterBarInline/FilterBarInline.mdx
Comment (bigfishdesign13) [reply src/components/FilterBarInline/FilterBarInline.mdx:407]: Yes, it is a little laggy, but acceptable in this example.  I'm curious to see what happens as the number of input children increases.  Let's keep an eye on it.

**Review by bigfishdesign13 — APPROVED** [review]
Looks great!  Let's just keep an eye on the potential lagginess.

**Pull request merged**
